internal/xds/translator/authentication.go

```diff
@@ -0,0 +1,430 @@
+		TransportSocket:      tSocket,
+	}, nil
+}
+
+func buildXdsUpstreamTLSSocket() (*core.TransportSocket, error) {
+	tlsCtx := &tls.UpstreamTlsContext{
+		CommonTlsContext: &tls.CommonTlsContext{
+			ValidationContextType: &tls.CommonTlsContext_ValidationContext{
+				ValidationContext: &tls.CertificateValidationContext{
+					TrustedCa: &core.DataSource{
+						Specifier: &core.DataSource_Filename{
+							Filename: "/etc/ssl/certs/ca-certificates.crt",
+						},
+					},
+				},
+			},
+		},
+	}
+
+	tlsCtxAny, err := anypb.New(tlsCtx)
+	if err != nil {
+		return nil, err
+	}
+
+	return &core.TransportSocket{
+		Name: wellknown.TransportSocketTls,
+		ConfigType: &core.TransportSocket_TypedConfig{
+			TypedConfig: tlsCtxAny,
+		},
+	}, nil
+}
+
+// patchRouteWithJwtConfig patches the provided route with a JWT PerRouteConfig, if the
+// route doesn't contain it.
+func patchRouteWithJwtConfig(route *routev3.Route, irRoute *ir.HTTPRoute) error { //nolint:unparam
+	if route == nil {
+		return errors.New("xds route is nil")
+	}
+	if irRoute == nil {
+		return errors.New("ir route is nil")
+	}
+
+	cfg := route.GetTypedPerFilterConfig()
+	if _, ok := cfg[jwtAuthenFilter]; !ok {
+		if !routeContainsJwtAuthn(irRoute) {
+			return nil
+		}
+
+		jwtPerRouteCfgProto, err := anypb.New(buildJwtPerRouteConfig(irRoute))
+		if err != nil {
+			return err
+		}
+
+		if cfg == nil {
+			route.TypedPerFilterConfig = make(map[string]*anypb.Any)
+		}
+
+		route.TypedPerFilterConfig[jwtAuthenFilter] = jwtPerRouteCfgProto
+	}
+
+	return nil
+}
+
+// buildJwtPerRouteConfig returns a JWT PerRouteConfig using "providers" as
+// the requirement name.
+func buildJwtPerRouteConfig(irRoute *ir.HTTPRoute) *jwtext.PerRouteConfig {
+	if irRoute == nil || !routeContainsJwtAuthn(irRoute) {
+		return nil
+	}
+
+	for providerKey := range irRoute.RequestAuthentication.JWT.Providers {
+		providerParts := strings.Split(providerKey, "-")
+		routeParts := strings.Split(irRoute.Name, "-")
+		if len(routeParts) < 2 || len(providerParts) < 2 {
+			// Route name and jwt providers key should start in the form of namespace-name.
+			return nil
+		}
+
+		// Split off the ns/name for the route and provider for equality comparison.
+		routeNsName := fmt.Sprintf("%s-%s", routeParts[0], routeParts[1])
+		providerNsName := fmt.Sprintf("%s-%s", providerParts[0], providerParts[1])
+
+		if routeNsName == providerNsName {
+			return &jwtext.PerRouteConfig{
+				RequirementSpecifier: &jwtext.PerRouteConfig_RequirementName{
+					RequirementName: irRoute.Name,
+				},
+			}
+		}
+	}
+
+	return nil
+}
+
+type jwksCluster struct {
+	name      string
+	addresses []string
+	port      int
+}
+
+// newJwksCluster returns a jwksCluster from the provided provider.
+func newJwksCluster(provider *v1alpha1.JwtAuthenticationFilterProvider) (*jwksCluster, error) {
+	if provider == nil {
+		return nil, errors.New("nil provider")
+	}
+
+	u, err := url.Parse(provider.RemoteJWKS.URI)
```

Comment:
is this needed here, or can this be done in the API validation layer ? (trying to ensure validation is not duplicated)

Reply:
The URL does get parsed/validated as part of validating the xds IR (gatewayapi runner). However, the URL is parsed here to get the scheme, hostname, and port. These values are used to create the xDS cluster, etc.

Reply:
sg thanks, reg validation, we should consider it doing more in the gateway -api layer so we can signal the errorMsgs to the user via status